core/src/main/scala/kafka/server/RaftRequestHandler.scala

```diff
@@ -48,6 +59,51 @@ class RaftRequestHandler(networkChannel: KafkaNetworkChannel,
             requestBody,
             response => sendResponse(request, Some(response)))
 
+        case ApiKeys.API_VERSIONS =>
+          sendResponse(request, Option(ApiVersionsResponse.apiVersionsResponse(10, 2)))
+
+        case ApiKeys.METADATA =>
+          val topics = new MetadataResponseData.MetadataResponseTopicCollection
+          if (connectionState.hostInfo().isPresent) {
+            topics.add(new MetadataResponseTopic()
+              .setErrorCode(Errors.NONE.code)
+              .setName("__cluster_metadata")
+              .setIsInternal(true)
+              .setPartitions(Collections.singletonList(new MetadataResponsePartition()
+                .setErrorCode(Errors.NONE.code())
+                .setPartitionIndex(0)
+                .setLeaderId(connectionState.id().toInt))))
+          }
+
+          val brokers = new MetadataResponseData.MetadataResponseBrokerCollection
+
+          if (connectionState.hostInfo().isPresent) {
+            val socketAddress = connectionState.hostInfo().get().address
+            brokers.add(new MetadataResponseData.MetadataResponseBroker()
+              .setNodeId(connectionState.id().toInt)
+              .setHost(socketAddress.getHostName)
+              .setPort(socketAddress.getPort))
+          }
+
+          sendResponse(request, Option(new MetadataResponse(
+            new MetadataResponseData()
+              .setTopics(topics)
+              .setBrokers(brokers))))
+
+        case ApiKeys.PRODUCE =>
+
+          val produceRequest = request.body[ProduceRequest]
+          produceRequest.partitionRecordsOrFail()
+          counter.increment().whenComplete{ (_, exception) =>
```

Comment:
Hmm. I had assumed that we would be taking the records from the request and appending them to the log. Otherwise, the comparison with a normal topic partition is not very meaningful. Could we create something similar to the mock state machine which allows us to append arbitrary records?

The other thing we need to do is make sure the `acks` behavior is consistent. I think the behavior we have here is consistent with acks=1, but a more useful comparison is the acks=all case.